### ISSUE TITLE
[32B Dense] Comparison v1 vs. v2 run

### ISSUE THREAD

**Issue by tvosch** [OpenEuroLLM/Taskboard#412]
# Goal

Pick the run to continue for the rest of the 32B dense run: v1 or v2. The deciding point for now is at 86k iterations or 1.4T tokens of the total 10-15T. 

### v1 run

- FP8
- `lr: 3e-4`
- `z-loss 1e-4`
- `defer_embedding_wgrad_compute: true`
- At 64k loss diverged: `z-loss` was lowered to `1e-5` `QK-layernorm: weight decay`, `RMS-norm weight decay`, Fused CE and z-loss gradients in FP32. This run is called B1
- Also a 86k fork with `defer_embedding_wgrad_compute: false` only
- wandb: https://wandb.ai/openeurollm-project/oellm_32b_dense

### v2 run

- BF16 (22% slower than v1)
- `lr: 1.76e-4`
- `defer_embedding_wgrad_compute: false`
- no z-loss
- After 60k iteration `z-loss: 1e-4` and `CE fusion: false`
- wandb: https://wandb.ai/openeurollm-project/oellm_32b_dense_v2

# Approach

Since comparing the two runs on either loss, validation loss or downstream evaluation with different learning rates, we expect the lower learning rate to be favoured within the constant learning rate phase -> so lower loss, higher eval score. The higher learning rate (under the condition that the run is stable) is to outperform on downstream evaluation only after proper annealing.

Annealing can be expensive so we approximate annealing with [WSM](https://arxiv.org/abs/2507.17634). @geoalgo  already reproduced this on prelude [here](https://github.com/OpenEuroLLM/Taskboard/issues/406)

For simplicity we do annealing on 5 checkpoints uniformly (so averaging with equal weights) across checkpoints of 2000 iterations apart (78k–86k; 230k–238k).

Downstream evaluation is done with `oellm-eval`

**Comment by tvosch** [comment]
# Without annealing on equal compute budget:

## Results

**Equal compute, plain checkpoints** (group averages):

| | open-sci 12 | DCLM other | reasoning | belebele | global-MMLU |
|---|---|---|---|---|---|
| v2 + z-loss @86k | **0.7248** | **0.5708** | **0.3075** | **0.8476** | **0.5878** |
| B1 @104k | 0.6918 | 0.5515 | 0.2623 | 0.8203 | 0.5649 |

v2 wins all five, by +0.019 to +0.045. Held-out loss: v2 1.3465, fork 1.3786, B1 1.4933 (own stacks).

**Comment by tvosch** [comment]
# On the effect of merging itself

**Merging helps all branches, by different amounts**:

| | gain |
|---|---|
| v2 + z-loss @86k | +0.0238 [+0.0205, +0.0274] |
| defer-off fork @86k | +0.0166 [+0.0130, +0.0201] |
| B1 @86k | +0.0235 [+0.0189, +0.0278] |
| B1 @238k | +0.0302 [+0.0266, +0.0342] |

The spread in gain (0.006–0.013) is smaller than the gap between trunks, so merging does not change
the equal-compute conclusion. It does move v2 vs B1@238k in B1's favour: v2 leads 3 of 5 groups as
plain checkpoints, 1 of 5 once both are merged. Uniform weighting beat tail-weighted
(`minus-sqrt__reverse`) on validation loss, 1.3107 vs 1.3129.

**Comment by tvosch** [comment]
# Downstream evals for merging uniformly 5 checkpoints
 

<img width="3080" height="1520" alt="Image" src="https://github.com/user-attachments/assets/f1177a8e-b8b2-4d7d-97f1-83f39d788a64" />

**Comment by JeniaJitsev** [comment]
I suggest to test a more modern merging procedure, eg following this: https://docs.nvidia.com/nemotron/nightly/nemotron/super3/pretrain.html#checkpoint-merging (especially 1 - sqrt part, which is also experimentally shown to be more advantageous in original WSM paper )
